chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomStore.kt

```diff
@@ -1,163 +1,155 @@
-                    val knownRoom = roomsMap[newRoom.id]!!
-                    // we don't get unread count and members with this event
-                    newRoom = newRoom.copy(memberUserIds = knownRoom.memberUserIds)
-                    if (knownRoom.unreadCount != null) {
-                        // we only get unread counts in initial state for 1000 rooms
-                        newRoom = newRoom.copy(unreadCount = knownRoom.unreadCount)
-                    }
-                    this += newRoom
-                    listOf(UserSubscriptionEvent.RoomUpdatedEvent(newRoom))
+                val addedEvents = addedRooms.map { UserInternalEvent.AddedToRoom(this[it]!!) }
+                val removedEvents = removedRooms.map { UserInternalEvent.RemovedFromRoom(it) }
+                val updatedEvents = (changedRooms + changedReadStates).toSet().map {
+                    UserInternalEvent.RoomUpdated(this[it]!!)
                 }
-                is UserSubscriptionEvent.ReadStateUpdatedEvent -> {
-                    var room = roomsMap[event.readState.roomId]!!
-                    room = room.copy(unreadCount = event.readState.unreadCount)
-                    this += room
-
-                    // Returning a copy with no cursor so that the returned applied event is
-                    // not translated to NewReadCursor but to RoomUpdated ChatEvent.
-                    // That is the relevant event from the perspective of this store.
-                    // The ChatEvent.NewReadCursor will be translated
-                    // into if CursorStore updates applying ReadStateUpdatedEvent and
-                    // returning it with the cursor.
-                    val eventWithNoCursor = event.copy(event.readState.copy(cursor = null))
-                    listOf(eventWithNoCursor)
+                val membershipEvents = changedMembers.flatMap { (roomId, existing, new) ->
+                    (new - existing).map { UserInternalEvent.UserJoinedRoom(it, roomId) } +
+                            (existing - new).map { UserInternalEvent.UserLeftRoom(it, roomId) }
                 }
-                is UserSubscriptionEvent.RoomDeletedEvent ->
-                    listOf(event.also { this -= event.roomId })
-                is UserSubscriptionEvent.RemovedFromRoomEvent ->
-                    listOf(event.also { this -= event.roomId })
-                is UserSubscriptionEvent.UserJoinedRoomEvent -> {
-                    val room = roomsMap[event.roomId]!!
-                    this += room.withAddedMember(event.userId)
-                    listOf(event)
+
+                addedEvents + removedEvents + updatedEvents + membershipEvents
+            }
+            is UserSubscriptionEvent.AddedToRoomEvent -> {
+                this.rooms[event.room.id] = event.room
+                this.members[event.membership.roomId] = event.membership.userIds.toSet()
+                this.unreadCounts[event.readState.roomId] = event.readState.unreadCount
```

Comment:
All the `this.` prefixes are not needed. I don't think they improve signal-to-noise ratio.